doc/man/man3/ds_read.3.man
'\" rtp
.\" vim: ft=nroff sw=4 noet nocin nosi com=b\:.\\\" fo+=tcqlorn tw=77
.\" =========================================================================
.\"
.\" @(#) doc/man/man3/ds_read.3.man
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2008-2011  Monavacon Limited <http://www.monavacon.com/>
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\" Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to copy, distribute and/or modify this manual under
.\" the terms of the GNU Free Documentation License, Version 1.3 or any later
.\" version published by the Free Software Foundation; with no Invariant
.\" Sections, no Front-Cover Texts, and no Back-Cover Texts.  A copy of the
.\" license is included in the section entitled "GNU Free Documentation
.\" License".
.\"
.\" Permission to use, copy and distribute this manual without modification,
.\" for any purpose and without fee or royalty is hereby granted, provided
.\" that both the above copyright notice and this permission notice appears
.\" in all copies and that the name of OpenSS7 Corporation not be used in
.\" advertising or publicity pertaining to distribution of this documentation
.\" or its contents without specific, written prior permission.  OpenSS7
.\" Corporation makes no representation about the suitability of this manual
.\" for any purpose.  It is provided "as is" without express or implied
.\" warranty.
.\"
.\" Permission is granted to process this file through groff and print the
.\" results, provided the printed document carries a copying permission
.\" notice identical to this one except for the removal of this paragraph
.\" (this paragraph not being relevant to the printed manual).
.\"
.\" OPENSS7 CORPORATION DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS MANUAL
.\" INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
.\" PARTICULAR PURPOSE, NON-INFRINGEMENT, OR TITLE; THAT THE CONTENTS OF THE
.\" DOCUMENT ARE SUITABLE FOR ANY PURPOSE, OR THAT THE IMPLEMENTATION OF SUCH
.\" CONTENTS WILL NOT INFRINGE ON ANY THIRD PARTY PATENTS, COPYRIGHTS,
.\" TRADEMARKS OR OTHER RIGHTS.  IN NO EVENT SHALL OPENSS7 CORPORATION BE
.\" LIABLE FOR ANY DIRECT, INDIRECT, SPECIAL OR CONSEQUENTIAL DAMAGES OR ANY
.\" DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER
.\" IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
.\" OUT OF OR IN CONNECTION WITH ANY USE OF THIS DOCUMENT OR THE PERFORMANCE
.\" OR IMPLEMENTATION OF THE CONTENTS THEREOF.
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this manual
.\" page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from the
.\" use of the information contained herein.  The author(s) may not have
.\" taken the same level of care in the production of this manual, which is
.\" licensed free of charge, as they might when working professionally.  The
.\" author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by the
.\" source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
.\" behalf of the U.S. Government ("Government"), the following provisions
.\" apply to you.  If the Software is supplied by the Department of Defense
.\" ("DoD"), it is classified as "Commercial Computer Software" under
.\" paragraph 252.227-7014 of the DoD Supplement to the Federal Acquisition
.\" Regulations ("DFARS") (or any successor regulations) and the Government
.\" is acquiring only the license rights granted herein (the license rights
.\" customarily provided to non-Government users).  If the Software is
.\" supplied to any unit or agency of the Government other than DoD, it is
.\" classified as "Restricted Computer Software" and the Government's rights
.\" in the Software are defined in paragraph 52.227-19 of the Federal
.\" Acquisition Regulations ("FAR") (or any successor regulations) or, in the
.\" cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the
.\" FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database openss7.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
search-truncate 40
search-ignore CGIQOSTU
.R2
.so openss7.macros
.\"
.\"
.TH DS_READ 3 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ Library Calls"
.\"
.\"
.SH NAME
.B ds_read
\- query information on an entry by name
.\"
.\"
.SH SYNOPSIS
.PP
.B #include <xds.h>
.HP
.BI "DS_status ds_read(OM_private object " session ", OM_private_object " context ", OM_object " name ", OM_object " selection ", OM_private_object *" result_return ", OM_sint *" invoke_id_return );
.\"
.\"
.SH INTERFACE
.PP
XDS
.[
XDS
.]
.\"
.\"
.SH DESCRIPTION
.PP
.BR ds_read ()
is used to extract information from an explicitly named entry.  It can also be
used to verify a distinguished name.
.PP
The result of this operation can be abandonned.
.\"
.\"
.SH ARGUMENTS
.PP
.TP
.IR session "\0\fBSession\fP(Object(Session)).
The directory session against which this operation is performed.  This must be
a private object.
.TP
.IR context "\0\fBContext\fP(Object(Context)).
The directory context to be used for this operation.  Note that
.B size-Limit
does not apply to this operation.  This argument must be a private object or
the constant
.RB { DS_DEFAULT_CONTEXT }.
.TP
.IR name "\0\fBName\fP(Object(Name)).
The name of the target object entry.  Any alises in the name will be
dereferenced unles prohibited by the
.B Dont-Dereference-Aliases
service command.
.TP
.IR selection "\0\fBSelection\fP(Object(Entry-Information-Selection)).
Specifies what information from the entry is requested.  Information about no
attribtues, all attributes or just for a named set can be chosen.  Attribute
types are always returned, but the attribute values need not be.  The possible
values of this argument are set out in [?].
.\"
.\"
.SH RETURN
.PP
.TP
.BR Status "(Status)
Whether the read was completed or not, if used synchronously, or whether the
operation was initiated, if used asyncrhonously.
.TP
.BR Result "(Object(Read-Result))
Upon successful completion of a syncrhonous call, the result contains the
distinguished name of the target object, and a flag indicating whether the
result came from the original entry or a copy, as well as any requested
attribute types and values.  Attribute information is only returned if access
rights are sufficient.
.TP
.BR Invoke-ID "(Integer)
The
.B Invoke-ID
of an asyncrhonous directory operation.
.\"
.\"
.SH ERRORS
.PP
This function can return a System-Error or one of the following Library-Errors:
bad-argument, bad-attribute, bad-class, bad-context, bad-name, bad-session,
bad-workspace, miscellaneous, missing-type, too-many-operations.
.PP
The following Directory-Errors may be returned by the function (or by
.BR ds_receive_result (3)
when used asynchronoulsy):
Attribute-Error, Name-Error, Referral, Security-Error, Service-Error.
.PP
An Attribute-Error (no-such-attribute) is reported if an explicit list of
attributes is specified by the
.I selection
argument, but none of them are present in the entry.  This error is not
reported if any of the selected attributes are present.
.PP
A Security-Error (insuficcient-access-rights) will only be reported where
access rights preclude the reading of all requested attribute values.
.PP
This function can return a Communications-Error.
.PP
This function can return the error constant
.RB [ DS_NO_WORKSPACE ].
.\"
.\"
.\".SH USAGE
.\".PP
.\"
.\"
.\".SH NOTICES
.\".PP
.\"
.\"
.SH IMPLEMENTATION
.PP
.BR ds_read ()
is implemented on top of the more complicated OSMIS 3.0 interface.
.\"
.\"
.\".SH FILES
.\".PP
.\"
.\"
.SH BUGS
.PP
.BR ds_read ()
has no known bugs.
.\"
.\"
.SH SEE ALSO
.PP
.BR XDS (3),
.BR XOM (3),
.BR ds_abandon (3).
.\"
.\"
.SH COMPATIBILITY
.PP
This implementation is compatible with other implementations of the XDS.
.[
XDS
.]
.\"
.\"
.SH CONFORMANCE
.PP
XDS.
.[
XDS
.]
Conformance is validated using an
.BR autotest (1)
script.
.\"and the
.\".BR test-xds (8)
.\"test suite executable.
.\"
.\"
.SH HISTORY
.PP
.BR ds_read ()
was first part of the
.BR XDS .
.[
XDS
.]
.\"
.\"
.XX
.[
xom c607
.]
.\"
.\"
.[
$LIST$
.]
.TA


NEWS
OpenSS7 -- history of user visible changes.  2015-01-15
$Id: NEWS,v 1.1.2.25 2011-05-10 13:45:30 brian Exp $
Copyright (c) 2008-2015  Monavacon Limited. <http://www.monavacon.com/>
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2001  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
'OpenSS7' package.

Major changes for release openss7-1.1.7.20150115
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Source code release on GitHub.

Major changes for release openss7-1.1.7.20131209
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build corrections for RHEL 6.4.

   - Scripts for country boundary, shoreline and rivers database
     preparation.

   - Database for ANSI T1.101 point code assignments.

   - Suppress extransous /proc/1/comm error messages when testing for
     systemd.

   - Handle PUIAS to Springdale rename for PUIAS distribution.

Major changes for release openss7-1.1.7.20131123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Correction to ca-cert handling in RHEL 6 install scriptlet.

   - RHEL 6 build changes.

   - Set permissions correctly when applying kernel module patch.

   - Do not duplicate certs when including ca-cert.

   - Avoid conflicting shelll variable in RPM install scriplet.

   - Remove /proc/1/comm checks for systemd in module loader.

Major changes for release openss7-1.1.7.20130209
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Corrections to RPM spec file.

Major changes for release openss7-1.1.7.20130129
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Compensate for RHEL6 lack of c_rehash.

   - Do not create sysfs nodes: they do not work properly.

   - Use no foreground flag in strace.service file.

   - More error traces for strerr logger.

   - Remove static device names (the do not work either).

   - Changes to support weak kernel modules.

   - Clean up weak module support.

Major changes for release openss7-1.1.7.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Better Archlinux install script support.

Major changes for release openss7-1.1.6.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Remove use of kill_litter_super().

Major changes for release openss7-1.1.3.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Updates and corrections to weak kernel module builds.

   - Repo support and corrections for Archlinux packages.

Major changes for release openss7-1.1.3.20130123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Additional support for Archlinux.

   - Add udev rules for strace and strerr.

   - Add system unit files for strace and strerr.

   - Add repo support for Archlinux packages.

   - Build Archlinux packages better.

Major changes for release openss7-1.1.3.20130121
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build support for SLES 11.

   - Handle certificates better.

   - Get configuration files in the right places for SLES 11.

   - Support in RPM spec file for certificates on SLES 11.

   - Create a new OpenSS7 certificate.

   - Repository support for SLES 11.

   - RPM spec file reworked for new RPM release.

   - Better certificate handling for RHEL.

   - Additional build support for RHEL 6 with certificates.


isode/quipu/ds_read.c
/*****************************************************************************

 @(#) $RCSfile$ $Name$($Revision$) $Date$

 -----------------------------------------------------------------------------

 Copyright (c) 2001-2007  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2000  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU General Public License as published by the Free Software
 Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 details.

 You should have received a copy of the GNU General Public License along with
 this program.  If not, see <http://www.gnu.org/licenses/>, or write to the
 Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 -----------------------------------------------------------------------------

 Last Modified $Date$ by $Author$

 -----------------------------------------------------------------------------

 $Log$
 *****************************************************************************/

#ident "@(#) $RCSfile$ $Name$($Revision$) $Date$"

static char const ident[] = "$RCSfile$ $Name$($Revision$) $Date$";

/* ds_read.c - */

#ifndef lint
static char *rcsid =
    "Header: /xtel/isode/isode/quipu/RCS/ds_read.c,v 9.0 1992/06/16 12:34:01 isode Rel";
#endif

/*
 * Header: /xtel/isode/isode/quipu/RCS/ds_read.c,v 9.0 1992/06/16 12:34:01 isode Rel
 *
 *
 * Log: ds_read.c,v
 * Revision 9.0  1992/06/16  12:34:01  isode
 * Release 8.0
 *
 */

/*
 *                                NOTICE
 *
 *    Acquisition, use, and distribution of this module and related
 *    materials are subject to the restrictions of a license agreement.
 *    Consult the Preface in the User's Manual for the full terms of
 *    this agreement.
 *
 */

#include "quipu/util.h"
#include "quipu/entry.h"
#include "quipu/read.h"
#include "quipu/policy.h"
#include "quipu/connection.h"
#include "pepsy.h"
#include "quipu/DAS-types.h"

extern LLog *log_dsap;

Attr_Sequence eis_select();
Attr_Sequence dsa_eis_select();
extern Attr_Sequence entry_find_type();
static cant_use_cache();
static attribute_not_cache();
extern AttributeType at_control;
extern unsigned bind_policy;
extern unsigned strong_policy;
extern DN mydsadn;
extern struct di_block *di_alloc();

do_ds_read(arg, error, result, binddn, target, di_p, dsp, quipu_ctx, authtype)
	struct ds_read_arg *arg;
	struct ds_read_result *result;
	struct DSError *error;
	DN binddn;
	DN target;
	struct di_block **di_p;
	char dsp;
	char quipu_ctx;
	char authtype;
{
	Entry entryptr;
	int retval;

#ifdef NOTUSED
	int authenticated;
#endif
	DN realtarget;
	char authp;

	DLOG(log_dsap, LLOG_TRACE, ("ds_read"));

	if (!dsp)
		target = arg->rda_object;

	if (!dsp && dsa_read_control(arg, result))
		return (DS_OK);

	if (target == NULLDN) {
		/* can't read from the root */
		error->dse_type = DSE_NAMEERROR;
		error->ERR_NAME.DSE_na_problem = DSE_NA_NOSUCHOBJECT;
		error->ERR_NAME.DSE_na_matched = NULLDN;
		return (DS_ERROR_REMOTE);
	}

	switch (find_entry
		(target, &(arg->rda_common), binddn, NULLDNSEQ, FALSE, &(entryptr), error, di_p,
		 OP_READ)) {
	case DS_OK:
		/* Filled out entryptr - carry on */
		break;
	case DS_CONTINUE:
		/* Filled out di_p - what do we do with it ?? */
		return (DS_CONTINUE);

	case DS_X500_ERROR:
		/* Filled out error - what do we do with it ?? */
		return (DS_X500_ERROR);
	default:
		/* SCREAM */
		LLOG(log_dsap, LLOG_EXCEPTIONS, ("do_ds_read() - find_entry failed"));
		return (DS_ERROR_LOCAL);
	}

	realtarget = get_copy_dn(entryptr);

	/* User can be authenticated by using strong authentication for this operation, or by using 
	   the credntials from the bind. */

	if (!manager(binddn))
		authp = entryptr->e_authp ? entryptr->e_authp->ap_readandcompare : AP_SIMPLE;
	else
		authp = AP_SIMPLE;

#ifdef NOTUSED
	authenticated = 0;

	/* Credentials provided in the ds_bind */
	if ((bind_policy & POLICY_ACCESS_READ) && (!dsp) && (binddn != NULLDN))
		authenticated = 1;
#endif

	/* Strong authentication */
	/* If it's there, check it, even if you won't believe it anyway */
	if ((retval = check_security_parms((caddr_t) arg,
					   _ZReadArgumentDataDAS,
					   &_ZDAS_mod,
					   arg->rda_common.ca_security,
					   arg->rda_common.ca_sig, &binddn)) != 0) {
		error->dse_type = DSE_SECURITYERROR;
		error->ERR_SECURITY.DSE_sc_problem = retval;
		dn_free(realtarget);
		return (DS_ERROR_REMOTE);
	}
#ifdef NOTUSED
	if ((strong_policy & POLICY_ACCESS_READ) &&
	    (arg->rda_common.ca_sig != (struct signature *) 0))
		authenticated = 1;
#endif

	/* entry has got a full list of attributes, eventually select one required */
	if (check_acl((authtype % 3) >= authp ? binddn : NULLDN, ACL_READ,
		      entryptr->e_acl->ac_entry, realtarget) == NOTOK) {
		if (dsp
		    && (check_acl(binddn, ACL_READ, entryptr->e_acl->ac_entry, realtarget) == OK)) {
			error->dse_type = DSE_SECURITYERROR;
			error->ERR_SECURITY.DSE_sc_problem = DSE_SC_AUTHENTICATION;
			dn_free(realtarget);
			return (DS_ERROR_REMOTE);
		} else {
			error->dse_type = DSE_SECURITYERROR;
			error->ERR_SECURITY.DSE_sc_problem = DSE_SC_ACCESSRIGHTS;
			dn_free(realtarget);
			return (DS_ERROR_REMOTE);
		}
	}

	if (entryptr->e_dsainfo && need_pseudo_dsa(entryptr, arg)) {

		/* Its a 7.0 or better DSA. pseudo attributes asked for. special routing req'd */

		if (dn_cmp(realtarget, mydsadn) == 0) {
			/* Its me - generate result */

			if ((result->rdr_entry.ent_attr =
			     dsa_eis_select(arg->rda_eis, entryptr, dsp ? NULLDN : binddn,
					    quipu_ctx, realtarget)) != NULLATTR)
				goto out;

			if (arg->rda_eis.eis_allattributes || arg->rda_eis.eis_select == NULLATTR)
				goto out;

			error->dse_type = DSE_ATTRIBUTEERROR;
			error->ERR_ATTRIBUTE.DSE_at_name = get_copy_dn(entryptr);
			error->ERR_ATTRIBUTE.DSE_at_plist.DSE_at_what = DSE_AT_NOSUCHATTRIBUTE;
			error->ERR_ATTRIBUTE.DSE_at_plist.DSE_at_type =
			    AttrT_cpy(arg->rda_eis.eis_select->attr_type);
			error->ERR_ATTRIBUTE.DSE_at_plist.DSE_at_value = NULLAttrV;
			error->ERR_ATTRIBUTE.DSE_at_plist.dse_at_next = DSE_AT_NOPROBLEM;
			dn_free(realtarget);
			return (DS_ERROR_REMOTE);

		} else {
			/* make referral to DSA in this entry */
			(*di_p) = di_alloc();
			(*di_p)->di_type = DI_TASK;
			(*di_p)->di_dn = realtarget;
			(*di_p)->di_target = dn_cpy(realtarget);
			(*di_p)->di_reftype = RT_UNDEFINED;
			(*di_p)->di_rdn_resolved = CR_RDNRESOLVED_NOTDEFINED;
			(*di_p)->di_aliasedRDNs = CR_NOALIASEDRDNS;
			(*di_p)->di_entry = entryptr;
			entryptr->e_refcount++;
			(*di_p)->di_state = DI_COMPLETE;

			return DS_CONTINUE;
		}

	}

	if (cant_use_cache(entryptr, binddn, arg->rda_eis, realtarget)) {
		int res = referral_dsa_info(realtarget, NULLDNSEQ, FALSE, entryptr, error, di_p,
					    arg->rda_common.ca_servicecontrol.
					    svc_options & SVC_OPT_PREFERCHAIN);
		dn_free(realtarget);
		return res;
	}

	if (dsp && (eis_check(arg->rda_eis, entryptr, binddn) != OK)) {
		/* Can only send public things over DSP - but user is entitled to more */
		error->dse_type = DSE_SECURITYERROR;
		error->ERR_SECURITY.DSE_sc_problem = DSE_SC_AUTHENTICATION;
		dn_free(realtarget);
		return (DS_ERROR_REMOTE);
	}

	if ((result->rdr_entry.ent_attr =
	     eis_select(arg->rda_eis, entryptr, dsp ? NULLDN : binddn, quipu_ctx,
			realtarget)) == NULLATTR)
		if (!arg->rda_eis.eis_allattributes && arg->rda_eis.eis_select != NULLATTR) {
			error->dse_type = DSE_ATTRIBUTEERROR;
			error->ERR_ATTRIBUTE.DSE_at_name = get_copy_dn(entryptr);
			error->ERR_ATTRIBUTE.DSE_at_plist.DSE_at_what = DSE_AT_NOSUCHATTRIBUTE;
			error->ERR_ATTRIBUTE.DSE_at_plist.DSE_at_type =
			    AttrT_cpy(arg->rda_eis.eis_select->attr_type);
			error->ERR_ATTRIBUTE.DSE_at_plist.DSE_at_value = NULLAttrV;
			error->ERR_ATTRIBUTE.DSE_at_plist.dse_at_next = DSE_AT_NOPROBLEM;
			dn_free(realtarget);
			return (DS_ERROR_REMOTE);
		}

      out:;
	result->rdr_entry.ent_dn = realtarget;
	result->rdr_entry.ent_iscopy = entryptr->e_data;
	result->rdr_entry.ent_age = (time_t) 0;
	result->rdr_entry.ent_next = NULLENTRYINFO;
	result->rdr_common.cr_requestor = NULLDN;
	/* if no error and NOT SVC_OPT_DONTDEREFERENCEALIASES then */
	/* the alias will have been derefeferenced -signified by */
	/* NO_ERROR !!! */
	result->rdr_common.cr_aliasdereferenced = (error->dse_type == DSE_NOERROR) ? FALSE : TRUE;
	return (DS_OK);

}

static
cant_use_cache(ptr, dn, eis, target)
	Entry ptr;
	DN dn;
	EntryInfoSelection eis;
	DN target;
{
	register Attr_Sequence as;
	char dfltacl = FALSE;

	if (dn == NULLDN)
		return FALSE;

	if ((ptr->e_data == E_DATA_MASTER) || (ptr->e_data == E_TYPE_SLAVE) ||
	    (ptr->e_complete == FALSE))
		return FALSE;

	/* see if more than cached data is required */

	if (eis.eis_allattributes) {
		struct acl_attr *aa;
		struct oid_seq *oidptr;

		/* look for attr acl */
		/* see if any attributes use can see */

		if (check_acl(NULLDN, ACL_READ, ptr->e_acl->ac_default, target) == NOTOK)
			if (check_acl(dn, ACL_READ, ptr->e_acl->ac_default, target) == OK)
				return TRUE;

		if (ptr->e_acl->ac_attributes == NULLACL_ATTR)
			return FALSE;

		for (aa = ptr->e_acl->ac_attributes; aa != NULLACL_ATTR; aa = aa->aa_next)
			for (oidptr = aa->aa_types; oidptr != NULLOIDSEQ; oidptr = oidptr->oid_next)
				/* The attribute is in the attribute ACL list */
				/* Would a referral help the DUA ? */
				if (check_acl(NULLDN, ACL_READ, aa->aa_acl, target) == NOTOK)
					if (check_acl(dn, ACL_READ, aa->aa_acl, target) == OK)
						return TRUE;

	} else {
		/* for each attribute in eis.eis_select, see is user entitled to it. */

		if (check_acl(NULLDN, ACL_READ, ptr->e_acl->ac_default, target) == NOTOK)
			if (check_acl(dn, ACL_READ, ptr->e_acl->ac_default, target) == OK)
				dfltacl = TRUE;

		for (as = eis.eis_select; as != NULLATTR; as = as->attr_link) {
			if (entry_find_type(ptr, as->attr_type) == NULLATTR)
				if (attribute_not_cached
				    (ptr, dn, grab_oid(as->attr_type), target, ACL_READ, dfltacl))
					return TRUE;

		}
	}
	return FALSE;
}

static
attribute_not_cached(ptr, dn, at, target, level, dfltacl)
	Entry ptr;
	DN dn;
	OID at;
	DN target;
	int level;
	char dfltacl;
{
	register struct acl_attr *aa;
	register struct oid_seq *oidptr;

	/* see if more than cached data is required */
	if (ptr->e_acl->ac_attributes == NULLACL_ATTR)
		return (dfltacl);

	for (aa = ptr->e_acl->ac_attributes; aa != NULLACL_ATTR; aa = aa->aa_next)
		for (oidptr = aa->aa_types; oidptr != NULLOIDSEQ; oidptr = oidptr->oid_next)
			if (oid_cmp(oidptr->oid_oid, at) == 0) {
				/* The attribute is in the attribute ACL list */
				/* Would a referral help the DUA ? */
				if (check_acl(NULLDN, level, aa->aa_acl, target) == NOTOK)
					if (check_acl(dn, level, aa->aa_acl, target) == OK)
						return TRUE;
				return FALSE;
			}
	return (dfltacl);

}

static Attr_Sequence
dsa_control_info()
{
	extern int slave_edbs;
	extern int master_edbs;
	extern int local_master_size;
	extern int local_slave_size;
	extern int local_cache_size;
	char buffer[LINESIZE];
	Attr_Sequence as;

	(void) sprintf(buffer,
		       "%d Master entries (in %d EDBs), %d Slave entries (in %d EDBs), %d Cached entries",
		       local_master_size, master_edbs, local_slave_size, slave_edbs,
		       local_cache_size);

	as = as_comp_alloc();
	as->attr_acl = NULLACL_INFO;
	as->attr_type = at_control;
	as->attr_link = NULLATTR;
	if ((as->attr_value = str2avs(buffer, as->attr_type)) == NULLAV) {
		as_free(as);
		return (NULLATTR);
	}

	return (as);
}

dsa_read_control(arg, result)
	struct ds_read_arg *arg;
	struct ds_read_result *result;
{

	if ((arg->rda_eis.eis_allattributes) ||
	    (arg->rda_eis.eis_infotypes == EIS_ATTRIBUTETYPESONLY))
		return FALSE;

	if ((arg->rda_eis.eis_select == NULLATTR)
	    || (arg->rda_eis.eis_select->attr_link != NULLATTR))
		return FALSE;

	if (AttrT_cmp(at_control, arg->rda_eis.eis_select->attr_type) != 0)
		return FALSE;

	if ((result->rdr_entry.ent_attr = dsa_control_info()) == NULLATTR)
		return FALSE;

	/* Fiddle DN - for DUA caching !!! */
	result->rdr_entry.ent_dn = dn_cpy(mydsadn);

	result->rdr_entry.ent_iscopy = FALSE;
	result->rdr_entry.ent_age = (time_t) 0;
	result->rdr_entry.ent_next = NULLENTRYINFO;
	result->rdr_common.cr_requestor = NULLDN;
	result->rdr_common.cr_aliasdereferenced = FALSE;

	return TRUE;
}

need_pseudo_dsa(eptr, arg)
	Entry eptr;
	struct ds_read_arg *arg;
{
	Attr_Sequence as;

	if (quipu_ctx_supported(eptr) <= 2)
		return FALSE;

	if (!quipu_version_7(eptr))
		return FALSE;

	if ((arg->rda_common.ca_servicecontrol.svc_options & SVC_OPT_DONTUSECOPY) != 0)
		return TRUE;

	if (arg->rda_eis.eis_allattributes)
		return FALSE;

	for (as = arg->rda_eis.eis_select; as != NULLATTR; as = as->attr_link) {
		if (check_avs_schema(as->attr_type, eptr->e_oc) != OK)
			return TRUE;
	}

	return FALSE;
}
